core_matcher.go

```diff
@@ -19,34 +19,48 @@ import (
 )
 
 // coreMatcher uses a finite automaton to implement the matchesForJSONEvent and MatchesForFields functions.
-// state is the start of the automaton
-// namesUsed is a map of field names that are used in any of the patterns that this automaton encodes. Typically,
-// patterns only consider a subset of the fields in an incoming data object, and there is no reason to consider
-// fields that do not appear in patterns when using the automaton for matching
-// the updateable fields are grouped into the coreStart member so they can be updated atomically using atomic.Load()
+// The updateable fields are grouped into the coreFields member so they can be updated atomically using atomic.Load()
 // and atomic.Store(). This is necessary for coreMatcher to be thread-safe.
 type coreMatcher struct {
-	updateable atomic.Value // always holds a *coreStart
+	updateable atomic.Value // always holds a *coreFields
 	lock       sync.Mutex
```

Comment:
do we still need this lock then?

Reply:
Yeah, but I should put in a comment explaining what's up.  The lock is so that only one thread can update the state machines at a time. But there can be many threads matching events with read-only access in concurrently with the updating thread, and the atomic.Value is so that the updating thread can update the maps and slices while the other threads are reading them.

Reply:
Got it. To me a RWMutex would have made it more clear then perhaps, depending on the runtime behavior though.

How long and often do you expect the state machine to be locked compared to reads? E.g., in `etcd` this type of MVCC is implemented using RWMutex.

Of course, if a write mutex is required for a long period, this is going to block readers on the whole object.

On the other hand, with just the atomic value in the struct I'm struggling to understand the benefit of the additional lock then.

Perhaps this one is worth a benchmark comparison? Even if both implementations are close, readability and following the code is easier using one type of atomic operation (e.g. RWMutex) vs atomic and Mutex types combined. At least for me it's harder to understand what's going on with these two types of atomic operations in the same struct.

Reply:
I have benchmarked this one quite a bit.  I'm super-impressed at the performance of atomic.Value - if there is no updating going on, the performance cost is very close to zero. And if I run an updating thread, the slow-down is linear and moderate. See concurrency_test.go.   I had another test there that steadily increased the frequency of updates but it didn't show anything surprising. 

Fortunately each individual addPattern is pretty short  so if you have a bunch of threads trying to put in their own updates it still works pretty well.  I did some reading and found that the discussion of RWMutex in the community discussions was full of caveats and gotchas, so for the moment I like the current lock & atomic.Value design.